syntax = "proto3";

option objc_class_prefix = "MSG";

package messages;

service MessageStream {
  rpc SubscribeToChannel (SubscriptionRequest) returns (stream Message) {}
  rpc PushToChannel (Push) returns (PushOkResponse) {}
  rpc CreateQueue (CreateQueueRequest) returns (CreateQueueOkResponse) {}
  rpc DeleteQueue (DeleteQueueRequest) returns (DeleteQueueOkResponse) {}
  rpc CreateChannel (CreateChannelRequest) returns (CreateChannelResponse) {}
}

message SubscriptionRequest {
  string channel_id = 1;
}

message Message {
  string id = 1;
  bytes content = 2;
  string topic = 3;
}

message Push {
  bytes content = 1;
  string topic = 2;
  string queue_label = 3;
}

message PushOkResponse {}

message CreateQueueRequest {
  string queue_label = 1;
}

message CreateQueueOkResponse {}

message DeleteQueueRequest {
  string queue_label = 1;
}

message DeleteQueueOkResponse {}

message CreateChannelRequest {
  string queue_label = 1;
  string topic = 2;
}

message CreateChannelResponse {
  string channel_id = 1;
}
